R/bootstrap.R

```diff
@@ -32,55 +36,113 @@ bootstrapPage <- function(..., title = NULL, responsive = NULL, theme = NULL) {
+    # that should be fine considering that, DOM tree order is preorder,
+    # depth-first traversal, and at least in the bootstrapPage(theme) case, we
+    # have control over the relative ordering.
+    # https://dom.spec.whatwg.org/#concept-tree
+    # https://stackoverflow.com/a/16113998/1583084
+    #
+    # Note also that since this is shinyOptions() (and not options()), the
+    # option is automatically reset when the app (or session) exits
+    if (isRunning()) {
+      shinyOptions(bootstrapTheme = theme)
+    } else {
+      # Technically, this a potential issue (someone trying to execute/render
+      # bootstrapLib outside of a Shiny app), but it seems that, in that case,
+      # you likely have other problems, since sliderInput() et al. already assume
+      # that Shiny is the one doing the rendering
+      #warning(
+      #  "It appears `shiny::bootstrapLib()` was rendered outside of an Shiny ",
+      #  "application context, likely by calling `as.tags()`, `as.character()`, ",
+      #  "or `print()` directly on `bootstrapLib()` or UI components that may ",
+      #  "depend on it (e.g., `fluidPage()`, etc). For 'themable' UI components ",
+      #  "(e.g., `sliderInput()`, `selectInput()`, `dateInput()`, etc) to style ",
+      #  "themselves based on the Bootstrap theme, make sure `bootstrapLib()` is ",
+      #  "provided directly to the UI and that the UI is provided direction to ",
+      #  "`shinyApp()` (or `runApp()`)", call. = FALSE
+      #)
+    }
 
-  bootstraplib::bootstrap()
+    bootstraplib::bs_dependencies(theme)
+  })
+}
+
+#' Obtain Shiny's Bootstrap Sass theme
+#'
+#' Intended for use by Shiny developers to create Shiny bindings with intelligent
+#' styling based on the [bootstrapLib()]'s `theme` value.
+#'
+#' @return If called at render-time (i.e., inside a [htmltools::tagFunction()]),
+#' and [bootstrapLib()]'s `theme` has been set to a [bootstraplib::bs_theme()]
+#' object, then this returns the `theme`. Otherwise, this returns `NULL`.
+#' @seealso [getCurrentOutputInfo()], [bootstrapLib()], [htmltools::tagFunction()]
+#' @examples
+#'
+#' library(bootstraplib)
+#' myThemableContent <- htmltools::tagFunction(function() {
+#'   theme <- getCurrentTheme()
+#'   if (is_bs_theme(theme)) {
+#'     tagList(
+#'       tags$head(
+#'         tags$style(HTML(
+#'           bs_sass(".text-50 { color: mix($body-bg, $body-color, 50%); }", theme)
+#'         ))
+#'       ),
+#'       h3(
+#'         "This color is 50% background and 50% foreground color",
+#'         class = "text-50"
+#'       )
+#'     )
+#'   } else {
+#'     "Try setting a theme = bs_theme()"
+#'   }
+#' })
+#'
+#' if (interactive()) {
+#'   ui <- fluidPage(
+#'     theme = bs_theme(bg = "yellow", fg = "red"),
+#'     myThemableContent
+#'   )
+#'   shinyApp(ui, function(input, output) {})
+#' }
+#'
+#' @export
+getCurrentTheme <- function() {
+  getShinyOption("bootstrapTheme")
```

Comment:
I wonder if, managing this through a session object that's a reactive value (similar to how `getCurrentOutputInfo()` works) would help us somehow solve the real-time theming problem